File: frontend/src/beta/ecs/AbilityEntitySystem.js
import {
    FIREBALL_DAMAGE,
    GUN_RANGE,
} from "../combat/Moves.js";
import {
    STUN_DAMAGE,
} from "../combat/Abilities.js";
import {
    PROTOTYPE_ABILITY_STATS,
    PROTOTYPE_ACTION_TO_ABILITY,
} from "../loadout/BotLoadout.js";
import { angleDelta, clamp, normalizeAngle, rayIntersectsCircle, segmentIntersectsCircle } from "../combat/geometry.js";
import { runEntityWorld, withComponentState } from "./EntityWorld.js";
import { abilityContract, EFFECT_TYPES } from "../combat/AbilityContracts.js";
import { resolveShieldInteraction } from "../combat/ShieldSystem.js";
import { ignoresHostileEffects } from "../combat/DefensiveState.js";

const ENTITY_TYPES = new Set([
    "proximityMine", "mineExplosion", "orbitalMarker", "orbitalExplosion",
    "gravityField", "gravityExplosion", "nullZone", "hunterDrone", "silenceWave", "temporalRewindZone",
]);

export function isAbilityEntity(entity) {
    return ENTITY_TYPES.has(entity?.type);
}

/**
 * Advances persistent ability entities through deterministic ordered systems.
 * Combat math is injected so this system does not own fighter rules.
 */
export function tickAbilityEntityWorld(world, combat) {
    return runEntityWorld({
        ...world,
        fighters: world.fighters.map((fighter) => ({ ...fighter, nullZoneSilenced: false })),
    }, [
        markMinesHitByAttacks(),
        tickMines(combat),
        tickNonMineEntities(combat),
    ]);
}

function markMinesHitByAttacks() {
    return (world) => ({
        entities: world.entities.map((entity) => entity.type === "proximityMine"
            && mineHitByCurrentAttack(entity, world)
            ? withComponentState(entity, { hitTriggered: true })
            : entity),
    });
}

function tickMines(combat) {
    return (world) => {
        let fighters = world.fighters.map((fighter) => ({ ...fighter }));
        const mines = world.entities.filter((entity) => entity.type === "proximityMine").map((mine) => {
            const traveled = Number(mine.traveled ?? 0);
            const moving = traveled < 176;
            return moving
                ? withComponentState(mine, {
                    x: clamp(mine.x + mine.velocityX, 12, world.width - 12),
                    y: clamp(mine.y + mine.velocityY, 12, world.height - 12),
                    traveled: traveled + Math.hypot(mine.velocityX, mine.velocityY),
                    ageMs: Number(mine.ageMs ?? 0) + world.stepMs,
                    armed: false,
                })
                : withComponentState(mine, {
                    velocityX: 0,
                    velocityY: 0,
                    ageMs: Number(mine.ageMs ?? 0) + world.stepMs,
                    armed: true,
                });
        });
        const triggered = new Set(mines.filter((mine) => (
            Number(mine.ageMs) >= 20_000
            || mine.hitTriggered
            || (mine.armed && fighters.some((fighter) => fighter.slot !== mine.ownerSlot
                && Math.hypot(fighter.x - mine.x, fighter.y - mine.y) <= 70 + fighter.size / 2))
        )).map((mine) => mine.id));
        let changed = true;
        while (changed) {
            changed = false;
            for (const source of mines.filter((mine) => triggered.has(mine.id))) {
                for (const target of mines) {
                    if (!triggered.has(target.id) && Math.hypot(target.x - source.x, target.y - source.y) <= 70 + target.size / 2) {
                        triggered.add(target.id);
                        changed = true;
                    }
                }
            }
        }
        const entities = world.entities.filter((entity) => entity.type !== "proximityMine");
        for (const mine of mines) {
            if (!triggered.has(mine.id)) {
                entities.push(mine);
                continue;
            }
            fighters = fighters.map((fighter) => {
                if (Math.hypot(fighter.x - mine.x, fighter.y - mine.y) > 70 + fighter.size / 2) return fighter;
                if (ignoresHostileEffects(fighter)) return fighter;
                const shield = resolveEntityShield(fighter, mine, "proximity_mine");
                return shield.preventedEffects.has(EFFECT_TYPES.DAMAGE) ? shield.fighter : combat.applyDamageToShape(shield.fighter, 18);
            });
            entities.push(withComponentState(mine, { id: `${mine.id}-blast`, type: "mineExplosion", size: 140, visibleMs: 300, spawnedThisTick: true }));
        }
        return { entities, fighters };
    };
}

function tickNonMineEntities(combat) {
    return (world) => {
        let fighters = world.fighters;
        const entities = [];
        for (const entity of world.entities) {
            if (entity.spawnedThisTick) {
                const readyNextTick = { ...entity };
                delete readyNextTick.spawnedThisTick;
                entities.push(readyNextTick);
                continue;
            }
            if (entity.type === "proximityMine") {
                entities.push(entity);
                continue;
            }
            const result = tickEntity(entity, { ...world, fighters }, combat);
            fighters = result.fighters;
            if (result.entity) entities.push(result.entity);
        }
        return { entities, fighters };
    };
}

function tickEntity(entity, world, combat) {
    if (entity.type === "silenceWave") return tickSilenceWave(entity, world);
    if (entity.type === "gravityField" || entity.type === "nullZone") return tickField(entity, world, combat);
    if (entity.type === "hunterDrone") return tickHunterDrone(entity, world, combat);
    if (entity.type === "orbitalMarker") return tickOrbitalMarker(entity, world, combat);
    if (entity.type === "temporalRewindZone") {
        const remainingMs = Number(entity.remainingMs ?? 3000) - world.stepMs;
        return { fighters: world.fighters, entity: remainingMs > 0 ? withComponentState(entity, { remainingMs }) : null };
    }
    if (["mineExplosion", "orbitalExplosion", "gravityExplosion"].includes(entity.type)) {
        const visibleMs = Number(entity.visibleMs ?? 0) - world.stepMs;
        return { fighters: world.fighters, entity: visibleMs > 0 ? withComponentState(entity, { visibleMs }) : null };
    }
    return { fighters: world.fighters, entity };
}

function tickSilenceWave(entity, world) {
    const start = { x: entity.x, y: entity.y };
    const end = {
        x: clamp(entity.x + entity.velocityX, 0, world.width),
        y: clamp(entity.y + entity.velocityY, 0, world.height),
    };
    const remainingMs = Number(entity.remainingMs ?? 1000) - world.stepMs;
    const hitSlots = [...(entity.hitSlots ?? [])];
    let blocked = false;
    const fighters = world.fighters.map((fighter) => {
        if (fighter.slot === entity.ownerSlot || hitSlots.includes(fighter.slot)
            || !segmentIntersectsCircle(start, end, { ...fighter, size: fighter.size + entity.size })) return fighter;
        if (ignoresHostileEffects(fighter)) return fighter;
        hitSlots.push(fighter.slot);
        const shield = resolveEntityShield(fighter, entity, "silence_pulse");
        if (shield.preventedEffects.has(EFFECT_TYPES.DEBUFF)) {
            blocked = true;
            return shield.fighter;
        }
        return { ...fighter, silencedMs: 2000, stunnedMs: Math.max(Number(fighter.stunnedMs ?? 0), 100), preparingAbility: null, preparingMs: 0 };
    });
    const hitEdge = end.x === 0 || end.x === world.width || end.y === 0 || end.y === world.height;
    return {
        fighters,
        entity: remainingMs > 0 && !hitEdge && !blocked ? withComponentState(entity, { ...end, remainingMs, hitSlots }) : null,
    };
}

function tickField(entity, world, combat) {
    const traveled = Number(entity.traveled ?? 0);
    const moving = traveled < 176;
    const ageMs = Number(entity.ageMs ?? 0) + world.stepMs;
    const fuseMs = entity.type === "gravityField" && !moving
        ? Math.max(0, Number(entity.fuseMs ?? 3000) - world.stepMs)
        : Number(entity.fuseMs ?? 0);
    const active = !moving && (entity.type !== "gravityField" || fuseMs <= 0);
    const remainingMs = Number(entity.remainingMs ?? (entity.type === "gravityField" ? 2000 : 5000)) - (active ? world.stepMs : 0);
    const field = moving
        ? withComponentState(entity, {
            x: clamp(entity.x + entity.velocityX, entity.size / 2, world.width - entity.size / 2),
            y: clamp(entity.y + entity.velocityY, entity.size / 2, world.height - entity.size / 2),
            traveled: traveled + Math.hypot(entity.velocityX, entity.velocityY), ageMs, armed: false,
        })
        : withComponentState(entity, { velocityX: 0, velocityY: 0, ageMs, fuseMs, remainingMs, armed: active });
    if (remainingMs <= 0) return { fighters: world.fighters, entity: null };
    let fighters = world.fighters;
    if (!moving && entity.type === "gravityField" && fuseMs > 0) {
        fighters = fighters.map((fighter) => {
            const dx = field.x - fighter.x;
            const dy = field.y - fighter.y;
            const distance = Math.hypot(dx, dy);
            if (distance <= 0.001 || distance > field.size / 2 + fighter.size / 2) return fighter;
            if (ignoresHostileEffects(fighter)) return fighter;
            return {
                ...fighter,
                x: clamp(fighter.x + dx / distance * 6, fighter.size / 2, world.width - fighter.size / 2),
                y: clamp(fighter.y + dy / distance * 6, fighter.size / 2, world.height - fighter.size / 2),
            };
        });
        return { fighters, entity: field };
    }
    if (active) {
        fighters = fighters.map((fighter) => {
            const dx = field.x - fighter.x;
            const dy = field.y - fighter.y;
            const distance = Math.hypot(dx, dy);
            if (distance > field.size / 2 + fighter.size / 2) return fighter;
            if (ignoresHostileEffects(fighter)) return fighter;
            if (entity.type === "nullZone") return { ...fighter, nullZoneSilenced: true };
            const stats = PROTOTYPE_ABILITY_STATS.gravity_grenade;
            const band = Math.min(3, Math.floor(distance / Math.max(1, Number(stats.radius) / 4)));
            const damage = Number(stats.maxDamage) - band * Number(stats.damageStep);
            if (field.damageApplied) return fighter;
            const shield = resolveEntityShield(fighter, field, "gravity_grenade");
            return shield.preventedEffects.has(EFFECT_TYPES.DAMAGE) ? shield.fighter : combat.applyDamageToShape(shield.fighter, damage);
        });
    }
    return {
        fighters,
        entity: entity.type === "gravityField" && active
            ? withComponentState(field, { id: `${field.id}-blast`, type: "gravityExplosion", visibleMs: 300, armed: true })
            : field,
    };
}

function tickHunterDrone(entity, world, combat) {
    const remainingMs = Number(entity.remainingMs ?? 6000) - world.stepMs;
    if (remainingMs <= 0) return { fighters: world.fighters, entity: null };
    const hp = Number(entity.hp ?? 50) - damageToDroneThisTick(entity, world, combat);
    if (hp <= 0) return { fighters: world.fighters, entity: null };
    let fighters = world.fighters;
        const target = fighters.filter((fighter) => fighter.slot !== entity.ownerSlot && Number(fighter.hp ?? 0) > 0)
        .sort((a, b) => Math.hypot(a.x - entity.x, a.y - entity.y) - Math.hypot(b.x - entity.x, b.y - entity.y))[0];
    let drone = withComponentState(entity, {
        hp, remainingMs, ageMs: Number(entity.ageMs ?? 0) + world.stepMs,
        shotCooldownMs: Math.max(0, Number(entity.shotCooldownMs ?? 0) - world.stepMs),
    });
    if (target) {
        const dx = target.x - drone.x;
        const dy = target.y - drone.y;
        const distance = Math.max(1, Math.hypot(dx, dy));
        const desiredRotation = Math.atan2(dy, dx) * 180 / Math.PI;
        const rotation = normalizeAngle(Number(drone.rotation ?? 0) + clamp(angleDelta(Number(drone.rotation ?? 0), desiredRotation), -8, 8));
        drone = withComponentState(drone, {
            x: clamp(drone.x + dx / distance * Math.min(4.5, distance), 14, world.width - 14),
            y: clamp(drone.y + dy / distance * Math.min(4.5, distance), 14, world.height - 14),
            rotation,
        });
        const shotRange = Number(PROTOTYPE_ABILITY_STATS.hunter_drone.range ?? 200);
        if (drone.shotCooldownMs <= 0 && rayIntersectsCircle(drone, rotation, shotRange, target)) {
            const ownerIndex = fighters.findIndex((fighter) => fighter.slot === entity.ownerSlot);
            const targetIndex = fighters.findIndex((fighter) => fighter.id === target.id);
            fighters = [...fighters];
            const shield = targetIndex >= 0 && !ignoresHostileEffects(fighters[targetIndex]) ? resolveEntityShield(fighters[targetIndex], drone, "hunter_drone") : { preventedEffects: new Set([EFFECT_TYPES.DAMAGE]) };
            if (shield.preventedEffects.has(EFFECT_TYPES.DAMAGE)) fighters[targetIndex] = shield.fighter;
            else if (targetIndex >= 0 && ownerIndex >= 0) [fighters[ownerIndex], fighters[targetIndex]] = combat.applyDamageFromShapes(fighters[ownerIndex], fighters[targetIndex], 3);
            else if (targetIndex >= 0) fighters[targetIndex] = combat.applyDamageToShape(fighters[targetIndex], 3);
            drone = withComponentState(drone, {
                shotCooldownMs: Number(PROTOTYPE_ABILITY_STATS.hunter_drone.shotCooldownMs ?? 1000),
                shotVisualMs: 120,
            });
        }
    }
    drone = withComponentState(drone, { shotVisualMs: Math.max(0, Number(drone.shotVisualMs ?? 0) - world.stepMs) });
    return { fighters, entity: drone };
}

function tickOrbitalMarker(entity, world, combat) {
    const fuseMs = Number(entity.fuseMs ?? 1500) - world.stepMs;
    if (fuseMs > 0) return { fighters: world.fighters, entity: withComponentState(entity, { fuseMs }) };
    const fighters = world.fighters.map((fighter) => {
        const distance = Math.hypot(fighter.x - entity.x, fighter.y - entity.y);
        if (distance > 130 + fighter.size / 2) return fighter;
        if (ignoresHostileEffects(fighter)) return fighter;
        const shield = resolveEntityShield(fighter, entity, "orbital_strike");
        return combat.applyDamageToShape(shield.fighter, 50 * Math.max(0.25, 1 - distance / 130));
    });
    return {
        fighters,
        entity: withComponentState(entity, { id: `${entity.id}-blast`, type: "orbitalExplosion", size: 260, visibleMs: 400 }),
    };
}

function resolveEntityShield(fighter, source, abilityId) {
    return resolveShieldInteraction(fighter, source, abilityContract(abilityId)?.shieldInteraction);
}

function mineHitByCurrentAttack(mine, world) {
    const { fighters, grenades = [], fireballs = [], entities } = world;
    if (fighters.some((fighter) => (fighter.entityHitIds ?? []).includes(mine.id))) return true;
    if ([...grenades, ...fireballs].some((entity) => Math.hypot(entity.x - mine.x, entity.y - mine.y) <= (Number(entity.size ?? 12) + mine.size) / 2)) return true;
    if (entities.some((entity) => entity.id !== mine.id && entity.type === "silenceWave"
        && Math.hypot(entity.x - mine.x, entity.y - mine.y) <= (Number(entity.size ?? 0) + mine.size) / 2)) return true;
    return fighters.some((fighter) => {
        if (fighter.gunShotActive && rayIntersectsCircle(fighter, Number(fighter.rotation ?? 0), GUN_RANGE, mine)) return true;
        const ability = PROTOTYPE_ACTION_TO_ABILITY[fighter.prototypeTriggered];
        const stats = PROTOTYPE_ABILITY_STATS[ability] ?? {};
        if (["pistol_shot", "concussive_shot", "rail_shot"].includes(ability)) return rayIntersectsCircle(fighter, Number(fighter.rotation ?? 0), Number(stats.range ?? 0), mine);
        if (["heavy_slash", "quick_jab", "thrust", "repulsor_burst"].includes(ability)) return Math.hypot(fighter.x - mine.x, fighter.y - mine.y) <= Number(stats.range ?? stats.radius ?? 0) + mine.size / 2;
        return false;
    });
}

function damageToDroneThisTick(drone, world, combat) {
    let damage = 0;
    for (const fighter of world.fighters) {
        if ((fighter.swingActiveMs ?? 0) > 0 && combat.isSwingHitting(fighter, drone)) damage += combat.incomingMeleeDamage(fighter, drone);
        if (fighter.gunShotActive && rayIntersectsCircle(fighter, Number(fighter.rotation ?? 0), Number(GUN_RANGE), drone)) damage += combat.incomingGunDamage(fighter, drone);
        if (fighter.stunCastActive && combat.stunHits(fighter, drone)) damage += STUN_DAMAGE * combat.attackerDamageMultiplier(fighter);
        const ability = PROTOTYPE_ACTION_TO_ABILITY[fighter.prototypeTriggered];
        const stats = PROTOTYPE_ABILITY_STATS[ability] ?? {};
        const distance = Math.hypot(fighter.x - drone.x, fighter.y - drone.y);
        const rayHit = ["pistol_shot", "concussive_shot", "rail_shot"].includes(ability)
            && rayIntersectsCircle(fighter, Number(fighter.rotation ?? 0), Number(stats.range ?? 0), drone);
        const areaHit = ["heavy_slash", "quick_jab", "thrust", "repulsor_burst", "phase_strike"].includes(ability)
            && distance <= Number(stats.range ?? stats.radius ?? 0) + drone.size / 2;
        if (rayHit || areaHit) damage += Number(stats.damage ?? 0) * combat.attackerDamageMultiplier(fighter);
    }
    for (const fireball of world.fireballs ?? []) if (fireball.type === "fireball" && combat.overlapsShape(fireball, drone)) damage += FIREBALL_DAMAGE * Number(fireball.damageMultiplier ?? 1);
    for (const grenade of world.grenades ?? []) if (grenade.type === "grenadeExplosion") damage += combat.grenadeDamageToFighter(grenade, drone);
    for (const effect of world.entities.filter((candidate) => !candidate.spawnedThisTick)) {
        const distance = Math.hypot(effect.x - drone.x, effect.y - drone.y);
        if (effect.type === "mineExplosion" && distance <= effect.size / 2 + drone.size / 2) damage += 18;
        if (effect.type === "gravityExplosion" && distance <= effect.size / 2 + drone.size / 2) damage += 35;
        if (effect.type === "orbitalExplosion" && distance <= effect.size / 2 + drone.size / 2) damage += 50 * Math.max(0.25, 1 - distance / 130);
    }
    return damage;
}

File: server/src/main/java/com/example/machiner/simulation/ecs/AbilityEntitySystem.java
package com.example.machiner.simulation.ecs;

import com.example.machiner.simulation.combat.AbilityContracts.EffectType;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/** Authoritative deterministic lifecycle and interaction system for ability entities. */
public final class AbilityEntitySystem {
    private AbilityEntitySystem() {}

    public record ShieldResult(boolean blocked, Set<EffectType> preventedEffects) {
        public static ShieldResult none() { return new ShieldResult(false, Set.of()); }
        public boolean prevents(EffectType effect) { return preventedEffects.contains(effect); }
    }

    public interface Combat<F extends AbilityEntityCombatant> {
        void damage(F fighter, int amount);
        void damageFromOwner(List<F> fighters, int ownerSlot, F target, int amount);
        int damageToEntity(ArenaEntity entity, List<F> fighters, List<ArenaEntity> entities);
        boolean entityHitByCurrentAttack(ArenaEntity entity, List<F> fighters, List<ArenaEntity> entities);
        default ShieldResult shield(F fighter, double sourceX, double sourceY, String abilityId) { return ShieldResult.none(); }
    }

    public static <F extends AbilityEntityCombatant> List<ArenaEntity> tick(
            List<ArenaEntity> entities,
            List<F> fighters,
            ArenaBounds arena,
            int stepMs,
            Combat<F> combat) {
        fighters.forEach(fighter -> fighter.setZoneSilenced(false));
        List<ArenaEntity> next = new ArrayList<>();
        tickTravelingAndPersistent(entities, fighters, arena, stepMs, combat, next);
        tickMines(entities, fighters, arena, stepMs, combat, next);
        tickMarkersAndEffects(entities, fighters, stepMs, combat, next);
        return next;
    }

    private static <F extends AbilityEntityCombatant> void tickTravelingAndPersistent(
            List<ArenaEntity> entities, List<F> fighters, ArenaBounds arena, int stepMs,
            Combat<F> combat, List<ArenaEntity> next) {
        for (ArenaEntity entity : entities) {
            if ("silenceWave".equals(entity.type())) {
                int remainingMs = entity.timerMs() - stepMs;
                double nextX = clamp(entity.x() + entity.velocityX(), 0, arena.width());
                double nextY = clamp(entity.y() + entity.velocityY(), 0, arena.height());
                boolean blocked = false;
                for (F fighter : fighters) {
                    if (fighter.entitySlot() == entity.ownerSlot()) continue;
                    if (segmentIntersectsCircle(entity.x(), entity.y(), nextX, nextY,
                            fighter.entityX(), fighter.entityY(), fighter.entitySize() / 2.0 + entity.size() / 2.0)) {
                        if (fighter.ignoresHostileEffects()) continue;
                        if (combat.shield(fighter, entity.x(), entity.y(), "silence_pulse").prevents(EffectType.DEBUFF)) {
                            blocked = true;
                            continue;
                        }
                        fighter.applySilence(2000);
                        fighter.applyStun(stepMs);
                        fighter.cancelPreparation();
                    }
                }
                boolean atEdge = nextX <= 0 || nextX >= arena.width() || nextY <= 0 || nextY >= arena.height();
                if (remainingMs > 0 && !atEdge && !blocked) next.add(new ArenaEntity(entity.id(), entity.type(), entity.ownerSlot(), nextX, nextY,
                        entity.size(), entity.velocityX(), entity.velocityY(), entity.traveled() + 150, remainingMs, true));
                continue;
            }
            if ("gravityField".equals(entity.type()) || "nullZone".equals(entity.type())) {
                tickField(entity, fighters, arena, stepMs, combat, next);
                continue;
            }
            if ("hunterDrone".equals(entity.type())) tickDrone(entity, entities, fighters, arena, stepMs, combat, next);
        }
    }

    private static <F extends AbilityEntityCombatant> void tickField(
            ArenaEntity entity, List<F> fighters, ArenaBounds arena, int stepMs,
            Combat<F> combat, List<ArenaEntity> next) {
        boolean moving = entity.traveled() < 176;
        int ageMs = entity.timerMs() + stepMs;
        double x = moving ? clamp(entity.x() + entity.velocityX(), entity.size() / 2.0, arena.width() - entity.size() / 2.0) : entity.x();
        double y = moving ? clamp(entity.y() + entity.velocityY(), entity.size() / 2.0, arena.height() - entity.size() / 2.0) : entity.y();
        double traveled = moving ? entity.traveled() + Math.hypot(entity.velocityX(), entity.velocityY()) : entity.traveled();
        boolean gravityDetonates = "gravityField".equals(entity.type()) && !moving && ageMs >= 3900;
        boolean armed = !moving && !"gravityField".equals(entity.type());
        int lifetimeMs = "gravityField".equals(entity.type()) ? 4000 : 5400;
        if (ageMs >= lifetimeMs) return;
        ArenaEntity field = new ArenaEntity(entity.id(), entity.type(), entity.ownerSlot(), x, y, entity.size(),
                moving ? entity.velocityX() : 0, moving ? entity.velocityY() : 0, traveled, ageMs, armed);
        if (!moving) for (F fighter : fighters) {
            double dx = x - fighter.entityX();
            double dy = y - fighter.entityY();
            double distance = Math.hypot(dx, dy);
            if (distance > entity.size() / 2.0 + fighter.entitySize() / 2.0) continue;
            if (fighter.ignoresHostileEffects()) continue;
            if ("nullZone".equals(entity.type())) fighter.setZoneSilenced(true);
            else if (!gravityDetonates && distance > 0.001) {
                fighter.setEntityPosition(
                        clamp(fighter.entityX() + dx / distance * 6, fighter.entitySize() / 2.0, arena.width() - fighter.entitySize() / 2.0),
                        clamp(fighter.entityY() + dy / distance * 6, fighter.entitySize() / 2.0, arena.height() - fighter.entitySize() / 2.0));
            } else if (gravityDetonates) {
                int band = Math.min(3, (int) Math.floor(distance / 30.0));
                if (!combat.shield(fighter, x, y, "gravity_grenade").prevents(EffectType.DAMAGE)) combat.damage(fighter, 35 - band * 5);
            }
        }
        if (!gravityDetonates) next.add(field);
        else next.add(new ArenaEntity(entity.id() + "-blast", "gravityExplosion", entity.ownerSlot(), x, y,
                entity.size(), 0, 0, 0, 300, true));
    }

    private static <F extends AbilityEntityCombatant> void tickDrone(
            ArenaEntity entity, List<ArenaEntity> entities, List<F> fighters, ArenaBounds arena, int stepMs,
            Combat<F> combat, List<ArenaEntity> next) {
        int ageMs = entity.timerMs() + stepMs;
        if (ageMs >= 6000) return;
        int hp = entity.hp() - combat.damageToEntity(entity, fighters, entities);
        if (hp <= 0) return;
        F target = fighters.stream().filter(fighter -> fighter.entitySlot() != entity.ownerSlot() && fighter.entityHp() > 0)
                .min(java.util.Comparator.comparingDouble(fighter -> Math.hypot(fighter.entityX() - entity.x(), fighter.entityY() - entity.y())))
                .orElse(null);
        double x = entity.x(), y = entity.y();
        if (target == null) {
            next.add(new ArenaEntity(entity.id(), entity.type(), entity.ownerSlot(), x, y, entity.size(),
                    entity.velocityX(), entity.velocityY(), entity.traveled(), ageMs, true, hp));
            return;
        }
        double dx = target.entityX() - x, dy = target.entityY() - y, distance = Math.max(1, Math.hypot(dx, dy));
        x = clamp(x + dx / distance * Math.min(4.5, distance), 14, arena.width() - 14);
        y = clamp(y + dy / distance * Math.min(4.5, distance), 14, arena.height() - 14);
        double desired = Math.toDegrees(Math.atan2(dy, dx));
        double current = Math.toDegrees(Math.atan2(entity.velocityY(), entity.velocityX()));
        double rotation = normalizeDegrees(current + clamp(angleDelta(current, desired), -8, 8));
        double directionX = Math.cos(Math.toRadians(rotation)), directionY = Math.sin(Math.toRadians(rotation));
        if (ageMs % 1000 < stepMs && rayIntersectsCircle(x, y, directionX, directionY, 200,
                target.entityX(), target.entityY(), target.entitySize() / 2.0)) {
            if (!target.ignoresHostileEffects()
                    && !combat.shield(target, x, y, "hunter_drone").prevents(EffectType.DAMAGE)) {
                combat.damageFromOwner(fighters, entity.ownerSlot(), target, 3);
            }
        }
        next.add(new ArenaEntity(entity.id(), entity.type(), entity.ownerSlot(), x, y, entity.size(),
                directionX, directionY, entity.traveled(), ageMs, true, hp));
    }

    private static <F extends AbilityEntityCombatant> void tickMines(
            List<ArenaEntity> entities, List<F> fighters, ArenaBounds arena, int stepMs,
            Combat<F> combat, List<ArenaEntity> next) {
        List<ArenaEntity> mines = entities.stream().filter(entity -> "proximityMine".equals(entity.type())).map(entity -> {
            boolean moving = entity.traveled() < 176;
            return moving
                    ? new ArenaEntity(entity.id(), entity.type(), entity.ownerSlot(),
                    clamp(entity.x() + entity.velocityX(), 12, arena.width() - 12),
                    clamp(entity.y() + entity.velocityY(), 12, arena.height() - 12), entity.size(),
                    entity.velocityX(), entity.velocityY(), entity.traveled() + Math.hypot(entity.velocityX(), entity.velocityY()), entity.timerMs() + stepMs, false)
                    : new ArenaEntity(entity.id(), entity.type(), entity.ownerSlot(), entity.x(), entity.y(), entity.size(),
                    0, 0, entity.traveled(), entity.timerMs() + stepMs, true);
        }).toList();
        Set<String> triggered = new HashSet<>();
        for (ArenaEntity mine : mines) {
            if (mine.timerMs() >= 20_000
                    || combat.entityHitByCurrentAttack(mine, fighters, entities)
                    || (mine.armed() && fighters.stream().anyMatch(fighter -> fighter.entitySlot() != mine.ownerSlot()
                    && Math.hypot(fighter.entityX() - mine.x(), fighter.entityY() - mine.y()) <= 70 + fighter.entitySize() / 2.0))) {
                triggered.add(mine.id());
            }
        }
        boolean changed;
        do {
            changed = false;
            for (ArenaEntity source : mines.stream().filter(mine -> triggered.contains(mine.id())).toList()) {
                for (ArenaEntity target : mines) {
                    if (!triggered.contains(target.id()) && Math.hypot(target.x() - source.x(), target.y() - source.y()) <= 70 + target.size() / 2.0) {
                        triggered.add(target.id());
                        changed = true;
                    }
                }
            }
        } while (changed);
        for (ArenaEntity mine : mines) {
            if (!triggered.contains(mine.id())) {
                next.add(mine);
                continue;
            }
            fighters.stream().filter(fighter -> Math.hypot(fighter.entityX() - mine.x(), fighter.entityY() - mine.y()) <= 70 + fighter.entitySize() / 2.0)
                    .forEach(fighter -> {
                        if (fighter.ignoresHostileEffects()) return;
                        if (!combat.shield(fighter, mine.x(), mine.y(), "proximity_mine").prevents(EffectType.DAMAGE)) combat.damage(fighter, 18);
                    });
            next.add(new ArenaEntity(mine.id() + "-blast", "mineExplosion", mine.ownerSlot(), mine.x(), mine.y(), 140, 0, 0, 0, 300, true));
        }
    }

    private static <F extends AbilityEntityCombatant> void tickMarkersAndEffects(
            List<ArenaEntity> entities, List<F> fighters, int stepMs, Combat<F> combat, List<ArenaEntity> next) {
        for (ArenaEntity entity : entities) {
            if (Set.of("proximityMine", "gravityField", "nullZone", "hunterDrone", "silenceWave").contains(entity.type())) continue;
            if ("orbitalMarker".equals(entity.type())) {
                int fuse = entity.timerMs() - stepMs;
                if (fuse > 0) next.add(new ArenaEntity(entity.id(), entity.type(), entity.ownerSlot(), entity.x(), entity.y(), entity.size(), 0, 0, 0, fuse, true));
                else {
                    fighters.forEach(fighter -> {
                        double distance = Math.hypot(fighter.entityX() - entity.x(), fighter.entityY() - entity.y());
                        if (distance <= 130 + fighter.entitySize() / 2.0 && !fighter.ignoresHostileEffects()) {
                            combat.shield(fighter, entity.x(), entity.y(), "orbital_strike");
                            combat.damage(fighter, (int) Math.round(50 * Math.max(0.25, 1 - distance / 130)));
                        }
                    });
                    next.add(new ArenaEntity(entity.id() + "-blast", "orbitalExplosion", entity.ownerSlot(), entity.x(), entity.y(), 260, 0, 0, 0, 400, true));
                }
            } else {
                int timer = entity.timerMs() - stepMs;
                if (timer > 0) next.add(new ArenaEntity(entity.id(), entity.type(), entity.ownerSlot(), entity.x(), entity.y(), entity.size(), 0, 0, 0, timer, true));
            }
        }
    }

    private static boolean segmentIntersectsCircle(double x1, double y1, double x2, double y2,
                                                   double cx, double cy, double radius) {
        double dx = x2 - x1, dy = y2 - y1;
        double lengthSquared = dx * dx + dy * dy;
        double t = lengthSquared <= 0 ? 0 : clamp(((cx - x1) * dx + (cy - y1) * dy) / lengthSquared, 0, 1);
        return Math.hypot(cx - (x1 + t * dx), cy - (y1 + t * dy)) <= radius;
    }

    private static boolean rayIntersectsCircle(double x, double y, double dx, double dy, double range,
                                               double cx, double cy, double radius) {
        double projection = (cx - x) * dx + (cy - y) * dy;
        if (projection < -radius || projection > range + radius) return false;
        double closestX = x + Math.max(0, Math.min(range, projection)) * dx;
        double closestY = y + Math.max(0, Math.min(range, projection)) * dy;
        return Math.hypot(cx - closestX, cy - closestY) <= radius;
    }

    private static double angleDelta(double from, double to) {
        return normalizeDegrees(to - from);
    }

    private static double normalizeDegrees(double degrees) {
        double normalized = degrees % 360;
        if (normalized > 180) normalized -= 360;
        if (normalized <= -180) normalized += 360;
        return normalized;
    }

    private static double clamp(double value, double min, double max) {
        return Math.max(min, Math.min(max, value));
    }
}
